code/go/internal/validator/semantic/validate_dashboards_used_by_value.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"path"
+
+	"github.com/pkg/errors"
+
+	ve "github.com/elastic/package-spec/code/go/internal/errors"
+	"github.com/elastic/package-spec/code/go/internal/fspath"
+	"github.com/elastic/package-spec/code/go/internal/pkgpath"
+)
+
+type reference struct {
+	ID   string
+	Name string
+	Type string
+}
+
+// ValidateVisualizationsUsedByValue warns if there are any Kibana
+// Dashboard that defines visualizations by reference instead of value.
+// That is, it warns if a Kibana dashbaord file, foo.json,
+// defines some visualization using reference (containing an element of
+// "visualization" type inside references key).
+func ValidateVisualizationsUsedByValue(fsys fspath.FS) ve.ValidationErrors {
+	var errs ve.ValidationErrors
+
+	filePaths := path.Join("kibana", "dashboard", "*.json")
+	objectFiles, err := pkgpath.Files(fsys, filePaths)
+	if err != nil {
+		errs = append(errs, errors.Wrap(err, "error finding Kibana Dashboard files"))
+		return errs
+	}
+
+	for _, objectFile := range objectFiles {
+		filePath := objectFile.Path()
+
+		objectReferences, err := objectFile.Values(`$.references[?(@.type=="visualization")]`)
```

Comment:
is this the right way to check for visualizations included by reference in a dashboard?

Reply:
I would want to confirm that this reference key in the dashboard json files is the one defining that a visualization is set as reference.

I've just tested that adding some visualizations from the library in the EPR integration, and it resulted in these changes:

```diff
--- packages/elastic_package_registry/kibana/dashboard/elastic_package_registry-313c2700-099b-11ed-91b6-3b1f9c2b2771.json
+++ packages/elastic_package_registry/kibana/dashboard/elastic_package_registry-313c2700-099b-11ed-91b6-3b1f9c2b2771.json
@@ -2215,6 +2215,38 @@
                 "title": "EPR Indexer - Get duration seconds per indexer - Percentile 95",
                 "type": "lens",
                 "version": "8.0.0"
+            },
+            {
+                "embeddableConfig": {
+                    "enhancements": {}
+                },
+                "gridData": {
+                    "h": 15,
+                    "i": "6beb6552-317f-442e-a6d5-397dc849aa98",
+                    "w": 24,
+                    "x": 0,
+                    "y": 82
+                },
+                "panelIndex": "6beb6552-317f-442e-a6d5-397dc849aa98",
+                "panelRefName": "panel_6beb6552-317f-442e-a6d5-397dc849aa98",
+                "type": "lens",
+                "version": "8.0.0"
+            },
+            {
+                "embeddableConfig": {
+                    "enhancements": {}
+                },
+                "gridData": {
+                    "h": 15,
+                    "i": "a3ae1ef0-3e98-4d2c-9b19-8b2ddc54ef4d",
+                    "w": 24,
+                    "x": 24,
+                    "y": 82
+                },
+                "panelIndex": "a3ae1ef0-3e98-4d2c-9b19-8b2ddc54ef4d",
+                "panelRefName": "panel_a3ae1ef0-3e98-4d2c-9b19-8b2ddc54ef4d",
+                "type": "visualization",
+                "version": "8.0.0"
             }
         ],
         "refreshInterval": {
@@ -2417,6 +2449,16 @@
             "id": "metrics-*",
             "name": "d237dd46-1074-47ee-8186-3957b956acd9:indexpattern-datasource-layer-cfdc23c6-0606-45ea-b1d2-782ba8103f82",
             "type": "index-pattern"
+        },
+        {
+            "id": "elastic_package_registry-b82baae0-1e06-11ed-92f0-1397dcee01a8",
+            "name": "6beb6552-317f-442e-a6d5-397dc849aa98:panel_6beb6552-317f-442e-a6d5-397dc849aa98",
+            "type": "lens"
+        },
+        {
+            "id": "elastic_package_registry-fd226440-1e06-11ed-92f0-1397dcee01a8",
+            "name": "a3ae1ef0-3e98-4d2c-9b19-8b2ddc54ef4d:panel_a3ae1ef0-3e98-4d2c-9b19-8b2ddc54ef4d",
+            "type": "visualization"
         }
     ],
     "type": "dashboard"
```

From this test, I've just realised that it should be checked whether or not there are references from both Lens and Visualization. Should it be checked for other types too ?